### PR TITLE
[WED-71] Amazon Product Advertising API Integration

### PR DESCRIPTION
**Pull request by ThePletch**
This also includes setting up Sidekiq/Redis.

Frontend work: The Request form should have a toggleable section that prompts for the item's name and category (e.g. "shoes") only if the provided link is not an Amazon link, and otherwise displays a notice saying "We'll fetch the name and category from Amazon".

Use the `AmazonProductApi.amazon_link?` method to see if a link is from Amazon.

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -276,10 +313,15 @@ DEPENDENCIES
...
   spring
   state_machines-activerecord
   therubyracer
   turbolinks
   uglifier (>= 1.3.0)
+  vacuum
   web-console (~> 2.0)
+
+BUNDLED WITH
+   1.11.2
```
Comment (ThePletch) [inline Gemfile.lock]: god damn it

File: .env.test [hunk 2]
```diff
@@ -6,3 +6,5 @@ WED_DB_USER=postgres
 WED_DB_PWD=The_Traub
 DEVISE_PEPPER=9431fb43f1cc76749ec8c3d6e044e5109637078a85daf99b173a6cf6a62198105c0f148a99876901fd124ee436044188d0594c4d467a07a4883d5476bfcd1892
 DEVISE_SECRET_KEY=abcdefghijklm123456789
+SIDEKIQ_USERNAME=async
```
Comment (ThePletch) [inline .env.test]: Note the lack of `REDIS_URL` here - this is intentional, as Sidekiq should never be hitting Redis during specs.

File: config/initializers/sidekiq.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  config.redis = { url: ENV['REDIS_URL'] }
+end
+
+Sidekiq.configure_client do |config|
+  config.redis = { url: ENV['REDIS_URL'] }
+end
+
+require 'sidekiq/web'
+Sidekiq::Web.use Rack::Auth::Basic do |username, password|
+  username == ENV['SIDEKIQ_USERNAME'] && password == ENV['SIDEKIQ_PASS']
```
Comment (ThePletch) [inline config/initializers/sidekiq.rb]: See infrastructure channel in Slack for login creds on production/staging.

File: Gemfile.lock (hunk 1 above)
Comment (caramcc) [inline Gemfile.lock]: bundler :new_moon_with_face: 

imo not worth it to spend time trying to fix this between systems, even if it's a bit of a nuisance that it keeps showing up, it doesn't really hurt anything to keep checking in / removing / etc

File: lib/amazon_product_api.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+      # if multiple results, return the first one, otherwise return result
+      results["Item"].kind_of?(Array) ? results["Item"][0] : results["Item"]
+    end
+
+    # adapted from http://stackoverflow.com/questions/1764605/scrape-asin-from-amazon-url-using-javascript
+    def asin_from_url(amazon_url)
+      asin_regex = /https?:\/\/www.amazon.com\/(?:[\w-]+\/)?(?:dp|gp\/product)\/(?:\w+\/)?(\w{10})/
+      if match = asin_regex.match(amazon_url)
+        match[1]
```
Comment (caramcc) [inline lib/amazon_product_api.rb]: Why `match[1]` and not `match[0]`?

Comment (ThePletch) [inline lib/amazon_product_api.rb]: It pulls out the ASIN's matched group `(\w{10})` rather than the whole URL.

Comment (caramcc) [inline lib/amazon_product_api.rb]: Makes sense. Yeah, a comment there would be helpful.

**Comment by caramcc** [comment]
:ok_hand: otherwise

**Pull request merged**